src/blockchain/state_transition_function.cpp

```diff
@@ -0,0 +1,274 @@
+    // other operations will get added as the functionality evolves
+    return outcome::success();
+  }
+
+  outcome::result<void> STF::processAttestations(
+      State &state, const std::vector<SignedVote> &attestations) const {
+    // get justifications, justified slots and historical block hashes are
+    // already upto date as per the processing in process_block_header
+    auto justifications = getJustifications(state);
+
+    // From 3sf-mini/consensus.py - apply votes
+    for (auto &signed_vote : attestations) {
+      auto &vote = signed_vote.data;
+      if (vote.source.slot >= state.historical_block_hashes.size()) {
+        return Error::INVALID_VOTE_SOURCE_SLOT;
+      }
+      if (vote.target.slot >= state.historical_block_hashes.size()) {
+        return Error::INVALID_VOTE_TARGET_SLOT;
+      }
+      // Ignore votes whose source is not already justified,
+      // or whose target is not in the history, or whose target is not a
+      // valid justifiable slot
+      if (not getBit(state.justified_slots.data(), vote.source.slot)
+          // This condition is missing in 3sf mini but has been added here
+          // because we don't want to re-introduce the target again for
+          // remaining votes if the slot is already justified and its tracking
+          // already cleared out from justifications map
+          or getBit(state.justified_slots.data(), vote.target.slot)
+          or vote.source.root
+                 != state.historical_block_hashes.data().at(vote.source.slot)
+          or vote.target.root
+                 != state.historical_block_hashes.data().at(vote.target.slot)
+          or vote.target.slot <= vote.source.slot
+          or not isJustifiableSlot(state.latest_finalized.slot,
+                                   vote.target.slot)) {
+        continue;
+      }
+
+      auto justifications_it = justifications.find(vote.target.root);
+      // Track attempts to justify new hashes
+      if (justifications_it == justifications.end()) {
+        justifications_it =
+            justifications.emplace(vote.target.root, std::vector<bool>{}).first;
+        justifications_it->second.resize(VALIDATOR_REGISTRY_LIMIT);
+      }
+
+      if (vote.validator_id >= justifications_it->second.size()) {
+        return Error::INVALID_VOTER;
+      }
+      justifications_it->second.at(vote.validator_id) = true;
+
+      size_t count = 0;
+      for (auto &&bit : justifications_it->second) {
+        if (bit) {
+          ++count;
+        }
+      }
+
+      // If 2/3 voted for the same new valid hash to justify
+      // in 3sf mini this is strict equality, but we have updated it to >=
+      // also have modified it from count >= (2 * state.config.num_validators)
+      // / 3 to prevent integer division which could lead to less than 2/3 of
+      // validators justifying specially if the num_validators is low in testing
+      // scenarios
+      if (3 * count >= 2 * state.config.num_validators) {
```

Comment:
Do you sure? Following 'Byzantine Generals Problem' quorum must be `>= 2/3 all + 1`